### ISSUE TITLE
Weapon setting "Disabled" turns into "Shield" in jpn-final

### ISSUE THREAD

**Issue by jonaeru** [fgsfdsfgs/perfect_dark#224]
Sorry, my modification may have something to do with it, but I'm not sure why. I have added the issue for now.
https://github.com/fgsfdsfgs/perfect_dark/pull/181

The 6th weapon setting in Challenge 30 is "Disabled", but the shield was actually set.
`シールド` (Shield) is wrong, `無効` (Disabled) is correct.

![image](https://github.com/fgsfdsfgs/perfect_dark/assets/64701608/cf6a89d6-6077-4299-9398-d7fd359166e8)

**Comment by BiatuAutMiahn** [comment]
Different versions have the different number of string definitions. Look at [Ryan Dwyer's decomp project](https://gitlab.com/ryandwyer/perfect-dark) in the strings section. He has a branch for exploding heads and I modified it to be an option. When doing so I had to add strings in a few different places and due to different regions, I found that some regions offset strings by like 2 indices.

Edit:
I didn't want to mess with existing offsets, so I just appended the string. I even had to bump the datatype from uint8_t to uint16_t to accommodate the indices.

**Comment by jonaeru** [comment]
Thank you. I have already modified the constants.
https://github.com/fgsfdsfgs/perfect_dark/pull/181/files

I had no problem with Weapon Set selection, but for some reason it happens with Challenge.
Maybe the index is shifted somewhere in the Japanese version.

**Comment by jonaeru** [comment]
Unrelated to this case, I should have also modified the following.

https://github.com/fgsfdsfgs/perfect_dark/blob/port/src/game/mplayer/mplayer.c#L1498-L1502

```
#if VERSION == VERSION_JPN_FINAL
	{ /*0x0b*/ L_MPWEAPONS_044, { WEAPON_TIMEDMINE,        WEAPON_CROSSBOW,    WEAPON_TIMEDMINE,   WEAPON_CROSSBOW,       WEAPON_MPSHIELD, WEAPON_DISABLED }, { MPFEATURE_WEAPON_CROSSBOW,        0,                           0,                              0                       }, WEAPON_TIMEDMINE,   WEAPON_TIMEDMINE,   WEAPON_TIMEDMINE, WEAPON_TIMEDMINE,      WEAPON_MPSHIELD, WEAPON_DISABLED }, // Close Combat
#else
	{ /*0x0b*/ L_MPWEAPONS_044, { WEAPON_COMBATKNIFE,      WEAPON_COMBATKNIFE, WEAPON_TIMEDMINE,   WEAPON_CROSSBOW,       WEAPON_MPSHIELD, WEAPON_DISABLED }, { MPFEATURE_WEAPON_CROSSBOW,        0,                           0,                              0                       }, WEAPON_COMBATKNIFE, WEAPON_COMBATKNIFE, WEAPON_TIMEDMINE, WEAPON_TIMEDMINE,      WEAPON_MPSHIELD, WEAPON_DISABLED }, // Close Combat
#endif
```

**Comment by jonaeru** [comment]
I tried changing weapon set in Challenge30 in `mpconfigs.c`, but it was not reflected.
https://github.com/fgsfdsfgs/perfect_dark/blob/port/src/mpconfigs.c#L1232-L1237
"Disabled" for Challenge30 is fixed at `0x26`?
Maybe the addition of the combat knife shifted one off and made it a shield? I am not sure...

I don't think the constant correction is a problem because it doesn't happen with Weapon Set.
MPconfig causes problems.

**Comment by fgsfdsfgs** [comment]
mpconfigs are loaded from the ROM. I'll need to patch them at load time to fix this. In fact that's already done, I just forgot to account for the JPN version.

**Comment by jonaeru** [comment]
I see, mystery solved. Thanks!

**Comment by jonaeru** [comment]
I have tried to fix it. I got the following error, so I fixed `constants.h`, but if there is a better way, please let me know.
https://github.com/fgsfdsfgs/perfect_dark/pull/379

```
# make -f Makefile.port -j4 ROMID=jpn-final
Host platform: i686-windows
Target platform: i686-windows
i686-w64-mingw32-gcc -std=c11 -c -DVERSION=5 -DVERSION_HASH=\"36f8494eb\" -DVERSION_TARGET=\"i686-windows\" -DVERSION_ROMID=\"jpn-final\" -DVERSION_BRANCH=\"enable-combatknife-jp\" -DMATCHING=0 -DPAL=0 -DPIRACYCHECKS=0 -DROM_SIZE=32 -D_LANGUAGE_C=1 -DAVOID_UB=1 -I include -I include/PR -I src/include -I src/generated/jpn-final -I src/lib/ultra/audio -I port/include -IC:/msys64/mingw32/include/SDL2 -g -fno-inline-functions -fno-strict-aliasing -funsigned-char -fwrapv -Wall -Wno-address -Wno-aggressive-loop-optimizations -Wno-int-in-bool-context -Wno-misleading-indentation -Wno-missing-braces -Wno-multichar -Wno-tautological-compare -Wno-unused-but-set-variable -Wno-unused-value -Wno-unused-variable -Wno-format-truncation -Wno-pointer-sign -Og -o build/jpn-final-port/port/src/preprocess.o port/src/preprocess.c
In file included from src/include/data.h:4,
                 from port/src/preprocess.c:9:
port/src/preprocess.c: In function 'preprocessMpConfigs':
src/include/constants.h:2973:68: warning: implicit declaration of function 'defined' [-Wimplicit-function-declaration]
 2973 | #define MPWEAPON_COMBATKNIFE      (VERSION == VERSION_JPN_FINAL && defined(PLATFORM_N64) ?    0 : 0x19)
      |                                                                    ^~~~~~~
port/src/preprocess.c:1054:61: note: in expansion of macro 'MPWEAPON_COMBATKNIFE'
 1054 |                         } else if (cfg->setup.weapons[j] >= MPWEAPON_COMBATKNIFE) {
      |                                                             ^~~~~~~~~~~~~~~~~~~~
src/include/constants.h:2973:76: error: 'PLATFORM_N64' undeclared (first use in this function); did you mean 'PLATFORM_X86'?
 2973 | #define MPWEAPON_COMBATKNIFE      (VERSION == VERSION_JPN_FINAL && defined(PLATFORM_N64) ?    0 : 0x19)
      |                                                                            ^~~~~~~~~~~~
port/src/preprocess.c:1054:61: note: in expansion of macro 'MPWEAPON_COMBATKNIFE'
 1054 |                         } else if (cfg->setup.weapons[j] >= MPWEAPON_COMBATKNIFE) {
      |                                                             ^~~~~~~~~~~~~~~~~~~~
src/include/constants.h:2973:76: note: each undeclared identifier is reported only once for each function it appears in
 2973 | #define MPWEAPON_COMBATKNIFE      (VERSION == VERSION_JPN_FINAL && defined(PLATFORM_N64) ?    0 : 0x19)
      |                                                                            ^~~~~~~~~~~~
port/src/preprocess.c:1054:61: note: in expansion of macro 'MPWEAPON_COMBATKNIFE'
 1054 |                         } else if (cfg->setup.weapons[j] >= MPWEAPON_COMBATKNIFE) {
      |                                                             ^~~~~~~~~~~~~~~~~~~~
make: *** [Makefile.port:314: build/jpn-final-port/port/src/preprocess.o] Error 1
```

**Comment by fgsfdsfgs** [comment]
This should be fixed in the latest commit.
